src/ReactiveDomain.Foundation/StreamStore/ReadModelBase.cs

```diff
@@ -9,48 +9,127 @@
+
+        /// <summary>
+        /// Creates a read model using the provided Functions to get a listener and reader.
+        /// If provided, the stream reader reads existing events before switching to the listener for live events.
+        /// </summary>
+        /// <param name="name">The name of the read model. Also used as the name of the listener and reader.</param>
+        /// <param name="getListener">A Func to get a new <see cref="IListener"/>.</param>
+        /// <param name="getReader">A Func to get a new <see cref="IStreamReader"/></param>
+        protected ReadModelBase(string name, Func<IListener> getListener, Func<IStreamReader> getReader = null) {
             Ensure.NotNull(getListener, nameof(getListener));
             _getListener = getListener;
+            _getReader = getReader;
             _listeners = new List<IListener>();
             _bus = new InMemoryBus($"{nameof(ReadModelBase)}:{name} bus",false);
             _queue = new QueuedHandler(_bus,$"{nameof(ReadModelBase)}:{name} queue");
             _queue.Start();
         }
-       private IListener AddNewListener() {
-           var l = _getListener();
-           lock (_listeners) {
-               _listeners.Add(l);
-           }
-           l.EventStream.SubscribeToAll(_queue);
-           return l;
-       }
+
+        private IListener AddNewListener() {
+            var l = _getListener();
+            lock (_listeners) {
+                _listeners.Add(l);
+            }
+            l.EventStream.SubscribeToAll(_queue);
+            return l;
+        }
+
+        /// <summary>
+        /// Get the positions of all listeners.
+        /// </summary>
+        /// <returns>A list of Tuples of listener names and checkpoints.</returns>
         public List<Tuple<string, long>> GetCheckpoint() {
             lock (_listeners) {
                 return _listeners.Select(l => new Tuple<string, long>(l.StreamName, l.Position)).ToList();
             }
         }
 
+        /// <summary>
+        /// The stream of events that handlers should subscribe to.
+        /// </summary>
         public ISubscriber EventStream => _bus;
 
+        /// <summary>
+        /// Start playback of a named stream.
+        /// </summary>
+        /// <param name="stream">The name of the stream to play back.</param>
+        /// <param name="checkpoint">The event to start with.</param>
+        /// <param name="blockUntilLive">If true, blocks returning from this method until the listener has caught up.</param>
+        /// <param name="cancelWaitToken">Cancellation token to cancel waiting if blockUntilLive is true.</param>
         public void Start(string stream, long? checkpoint = null, bool blockUntilLive = false, CancellationToken cancelWaitToken = default(CancellationToken)) {
-           
+            if (_getReader != null) {
+                using (var reader = _getReader()) {
+                    reader.EventStream.SubscribeToAll(_queue);
```

Comment:
Is SubscribeToAll correct here?

Reply:
Because `EventStream` is an `ISubscriber, that appears to be the only way to get events of any type. Under the covers, `EventSream` points to an `InMemoryBus` that would allow us to use `Subscribe(IMessageHandler handler, bool includeDerived)` or `SubscribeToAll()`, but that property is not public.